app/models/manageiq/providers/embedded_ansible/automation_manager/playbook_runner.rb

```diff
@@ -0,0 +1,116 @@
+class ManageIQ::Providers::EmbeddedAnsible::AutomationManager::PlaybookRunner < ::Job
+  # options are job table columns, including options column which is the playbook context info
+  def self.create_job(options)
+    super(name, options)
+  end
+
+  def start
+    if options[:inventory]
+      queue_signal(:create_job_template)
+    else
+      queue_signal(:create_inventory)
+    end
+  end
+
+  def create_inventory
+    set_status('creating inventory')
+    tower = playbook.manager
+    hosts = options[:hosts] || options[:extra_vars][:hosts]
+    inventory =
+      if hosts == 'localhost'
+        tower.provider.default_inventory
+      else
+        inventory_name = "#{playbook.name}_#{Time.zone.now.to_i}"
+        ManageIQ::Providers::EmbeddedAnsible::AutomationManager::Inventory.create_inventory_raw(tower, inventory_name, hosts)
+      end
+    options[:inventory] = inventory.id
+    save!
+    queue_signal(:create_job_template)
+  end
+
+  def create_job_template
+    set_status('creating job template')
+    raw_job_template = playbook.create_job_template_raw(options)
+    options[:job_template_ref] = raw_job_template.id
+    save!
+
+    queue_signal(:launch_ansible_tower_job)
+  end
+
+  def poll_ansible_tower_job_status
+    set_status('waiting for tower job to complete')
+    if tower_job.raw_status.completed?
+      queue_signal(:finish)
+    else
+      queue_signal(:poll_ansible_tower_job_status, :deliver_on => Time.now.utc + 1.minute)
```

Comment:
At a minimal, if no inventory is passed in - each Runner 'Run' will take 3 minutes. Is the 1 minute time here to guarantee the synchronous saves complete?

Reply:
No. This is to requeue itself if the tower job is not completed. And yes, if the playbook runs quick enough, the 1 minute waiting time seems too long. But what's the right interval for best user experience?